modules/xfem/src/efa/EFAFragment2D.C

```diff
@@ -334,345 +334,86 @@ EFAFragment2D::split()
     }
   }
-  else if (cut_nodes.size() == 2)
-  {
-    if ((cut_nodes[0] == 0 && cut_nodes[1] == 2) || (cut_nodes[0] == 2 && cut_nodes[1] == 0))
-    {
-      EFAFragment2D * new_frag1 = new EFAFragment2D(_host_elem, false, NULL);
-      EFAFragment2D * new_frag2 = new EFAFragment2D(_host_elem, false, NULL);
-      EFANode * node1 = _boundary_edges[0]->getNode(0);
-      EFANode * node2 = _boundary_edges[1]->getNode(0);
-      EFANode * node3 = _boundary_edges[2]->getNode(0);
-      EFANode * node4 = _boundary_edges[3]->getNode(0);
 
-      new_frag1->addEdge(new EFAEdge(node1, node2));
-      new_frag1->addEdge(new EFAEdge(node2, node3));
-      new_frag1->addEdge(new EFAEdge(node3, node1));
-
-      new_frag2->addEdge(new EFAEdge(node3, node4));
-      new_frag2->addEdge(new EFAEdge(node4, node1));
-      new_frag2->addEdge(new EFAEdge(node1, node3));
-
-      new_fragments.push_back(new_frag1);
-      new_fragments.push_back(new_frag2);
+  if ((edge_cut_count + node_cut_count) > 1) // any two cuts case
+  {
+    EFAFragment2D * new_frag0 = new EFAFragment2D(_host_elem, false, NULL);
+    EFAFragment2D * new_frag1 = new EFAFragment2D(_host_elem, false, NULL);
+    if (fragment_nodes[0].size() >= 3)
+    { // check to make sure an edge wasn't cut
+      for (unsigned int inode = 0; inode < fragment_nodes[0].size() - 1;
+           inode++) // assemble fragment 1 (0)
+        new_frag0->addEdge(new EFAEdge(fragment_nodes[0][inode], fragment_nodes[0][inode + 1]));
+
+      new_frag0->addEdge(
+          new EFAEdge(fragment_nodes[0][fragment_nodes[0].size() - 1], fragment_nodes[0][0]));
+
+      new_fragments.push_back(new_frag0);
     }
-    else if ((cut_nodes[0] == 1 && cut_nodes[1] == 3) || (cut_nodes[0] == 3 && cut_nodes[1] == 1))
-    {
-      EFAFragment2D * new_frag1 = new EFAFragment2D(_host_elem, false, NULL);
-      EFAFragment2D * new_frag2 = new EFAFragment2D(_host_elem, false, NULL);
-      EFANode * node1 = _boundary_edges[0]->getNode(0);
-      EFANode * node2 = _boundary_edges[1]->getNode(0);
-      EFANode * node3 = _boundary_edges[2]->getNode(0);
-      EFANode * node4 = _boundary_edges[3]->getNode(0);
 
-      new_frag1->addEdge(new EFAEdge(node1, node2));
-      new_frag1->addEdge(new EFAEdge(node2, node4));
-      new_frag1->addEdge(new EFAEdge(node4, node1));
+    if (fragment_nodes[1].size() >= 3)
+    { // check to make sure the cut wasn't straight on an edge
+      for (unsigned int inode = 0; inode < fragment_nodes[1].size() - 1;
+           inode++) // assemble fragment 2 (1)
+        new_frag1->addEdge(new EFAEdge(fragment_nodes[1][inode], fragment_nodes[1][inode + 1]));
 
-      new_frag2->addEdge(new EFAEdge(node2, node3));
-      new_frag2->addEdge(new EFAEdge(node3, node4));
-      new_frag2->addEdge(new EFAEdge(node4, node2));
+      new_frag1->addEdge(
+          new EFAEdge(fragment_nodes[1][fragment_nodes[1].size() - 1], fragment_nodes[1][0]));
```

Comment:
This block is the same code as the block above (just with 1 instead of 0) This could be made more concise (and maintainable with a two step for loop)

Reply:
Good point! I just pushed an update that cleans that up.